src/java/org/apache/cassandra/index/sai/disk/v2/sortedbytes/SortedBytesReader.java

```diff
@@ -0,0 +1,157 @@
+ */
+public class SortedBytesReader
+{
+    private final SortedBytesWriter.Meta meta;
+    private final FileHandle trieHandle;
+    private final BytesRef term;
+    private final LongValues offsets;
+    private long pointId = -1;
+
+    public SortedBytesReader(SortedBytesWriter.Meta meta,
+                             FileHandle trieHandle,
+                             IndexInput offsetsInput) throws IOException
+    {
+        this.meta = meta;
+        this.trieHandle = trieHandle;
+
+        ByteArrayIndexInput offsestMetaInput = new ByteArrayIndexInput("", meta.offsetMetaBytes);
+
+        DirectMonotonicReader.Meta offsetsMeta = DirectMonotonicReader.loadMeta(offsestMetaInput, meta.offsetBlockCount, DIRECT_MONOTONIC_BLOCK_SHIFT);
+        RandomAccessInput offsetSlice = offsetsInput.randomAccessSlice(0, offsetsInput.length());
+
+        offsets = DirectMonotonicReader.getInstance(offsetsMeta, offsetSlice);
+
+        term = new BytesRef(meta.maxTermLength);
+    }
+
+    /**
+     * Returns the point id / ordinal of the target term, it not matching the next greater.
+     * @param term target term to lookup
+     * @return point id / ordinal or the target
+     * @throws IOException
+     */
+    public long seekToBytes(ByteComparable term) throws IOException
+    {
+        try (TrieRangeIterator reader = new TrieRangeIterator(trieHandle.instantiateRebufferer(),
+                                                              meta.trieFP,
+                                                              term,
+                                                              null,
+                                                              true,
+                                                              true))
+        {
+            final Iterator<Pair<ByteSource, Long>> iterator = reader.iterator();
+
+            if (iterator.hasNext())
+            {
+                final Pair<ByteSource, Long> pair = iterator.next();
+                return pair.right.longValue();
+            }
+            return -1;
+        }
+    }
+
+    /**
+     * Look up the ByteComparable of a target point id
+     *
+     * @param target Point id to lookup
+     * @param bytesInput Bytes input stream
+     * @return ByteComparable at the target point id
+     * @throws IOException
+     */
+    public ByteComparable seekExact(long target, IndexInput bytesInput) throws IOException
+    {
+        if (target < 0 || target >= meta.count)
+        {
+            throw new IndexOutOfBoundsException();
+        }
+        final long blockIndex = target >>> TERMS_DICT_BLOCK_SHIFT;
+        final long blockAddress = offsets.get(blockIndex);
+        bytesInput.seek(blockAddress);
+        this.pointId = (blockIndex << TERMS_DICT_BLOCK_SHIFT) - 1;
+        BytesRef term = null;
+        do
+        {
+            term = next(bytesInput);
+        } while (this.pointId < target);
+        return fixedLength(term);
```

Comment:
This looks extremely unsafe. `fixedLength` returns a lambda that captures the reference to `term.bytes` bytearray. That array seems to be reused and overwritten when we call `next()`. Therefore calling `next()` invalidates (corrupts) the `ByteComparable` returned from this method. 

I'd prefer to:
- make a copy into caller provided byte array (safe)
- write a really big warning in capital letters that this API is expert-only unsafe and explain all the limitations

Reply:
Agree, I missed this one. Although the class is not thread safe, the seek methods are. We should be using `ByteRefBuilder` and `.toBytesRef` to avoid the threading issue and take the perf hit.

Reply:
There is at present no `IndexInput#clone` like Lucene, an independent `IndexInput` must be passed in.

Term issue fixed.